packages/api-client/src/api/serializers/product.ts

```diff
@@ -80,45 +110,71 @@ const buildBreadcrumbs = (included, product) => {
   name: product.attributes.name,
   slug: product.attributes.slug,
   sku: product.attributes.sku,
   optionTypes: deserializeOptionTypes(attachments, product),
-  optionValues: deserializeOptionValues(attachments, variant),
-  images: deserializeImages(attachments, defaultVariant, variant),
+  optionValues: deserializeOptionValues(attachments, mainVariant),
+  images: deserializeImages(attachments, mainVariant, fallbackVariant),
   breadcrumbs: buildBreadcrumbs(attachments, product),
   properties: deserializeProperties(attachments, product),
-  displayPrice: variant.attributes.display_price,
+  displayPrice: mainVariant.attributes.display_price,
   price: {
-    original: variant.attributes.price,
-    current: variant.attributes.price
+    original: mainVariant.attributes.price,
+    current: mainVariant.attributes.price
   },
-  inStock: variant.attributes.in_stock
+  inStock: mainVariant.attributes.in_stock
 });
 
-const findProductVariants = (product, included) =>
-  included.filter((e) => e.type === 'variant' && e.relationships.product.data.id === product.id);
-
 export const deserializeSingleProductVariants = (apiProduct) => {
-  const variants = findProductVariants(apiProduct.data, apiProduct.included);
-  const defaultVariant = variants.find((e) => e.attributes.is_master) || variants[0];
-  return variants.map((variant) => deserializeProductVariant(apiProduct.data, variant, defaultVariant, apiProduct.included));
-};
+  const attachments = apiProduct.included;
+  const productId = apiProduct.data.id;
+
+  const groupedVariants = groupIncluded(
+    attachments,
+    {
+      primaryVariants: (document) => (isVariantOfProduct(document, productId) && document.attributes.is_master),
+      optionVariants: (document) => (isVariantOfProduct(document, productId) && !document.attributes.is_master)
+    }
+  );
+  const primaryVariant = groupedVariants.primaryVariants[0];
 
-export const deserializeLimitedVariants = (apiProducts) =>
-  apiProducts.data.map((product) => {
-    const attachments = apiProducts.included;
-    const variants = findProductVariants(product, attachments);
-    const defaultVariant = variants.find((e) => e.attributes.is_master) || variants[0];
+  return groupedVariants.optionVariants.map(
+    (variant) => deserializeProductVariant(apiProduct.data, variant, primaryVariant, attachments)
+  );
+};
 
-    return deserializeProductVariant(product, defaultVariant, defaultVariant, attachments);
+export const deserializeLimitedVariants = (apiProducts) => {
+  const attachments = apiProducts.included;
+
+  return apiProducts.data.map((product) => {
+    const productId = product.id;
+    const defaultVariantId = product.relationships.default_variant.data.id;
+
+    const groupedVariants = groupIncluded(
+      attachments,
+      {
+        primaryVariants: (document) => (isVariantOfProduct(document, productId) && document.attributes.is_master),
+        defaultOptionVariants: (document) => (
```

Comment:
Do we want to take a random non-master variant as a fallback for images? Won't that cause the experience to be inconsistent, as we could pull images from a different variant after e.g. refreshing the page?

Reply:
Hm, but we're taking variant order here from the `included` array in the response, right? Will they be sorted?
I think sorting that explicitly by attributes.position won't hurt ;)